java/code/src/com/suse/manager/reactor/messaging/RegisterMinionEventMessageAction.java

```diff
@@ -127,69 +128,67 @@ public void execute(EventMessage msg) {
+     * @param startupGrains start up grains start_event_grains
+     * @return Machine Id
+     */
+    private Optional<String> getMachineId(String minionId, Optional<MinionStartupGrains> startupGrains) {
+        Optional<String> machineId =
+                Opt.or(startupGrains.flatMap(msg -> msg.getMachineId()), ()-> SALT_SERVICE.getMachineId(minionId));
+        if (!machineId.isPresent()) {
+            LOG.info("Cannot find machine id for minion: " + minionId);
+        }
+        return machineId;
+    }
 
     /**
      * Performs minion registration.
-     *
      * @param minionId minion id
      * @param isSaltSSH true if a salt-ssh system is bootstrapped
      * @param actKeyOverride label of activation key to be applied to the system.
-     *                              If left empty, activation key from grains will be used.
+     *                       If left empty, activation key from grains will be used.
+     * @param machineId Machine Id of the minion
      */
-    private void registerMinion(String minionId, boolean isSaltSSH,
-                                Optional<Long> saltSSHProxyId,
-                                Optional<String> actKeyOverride) {
-
-        // Match minions via their machine id
-        Opt.fold(SALT_SERVICE.getMachineId(minionId),
-            () -> {
-                LOG.info("Cannot find machine id for minion: " + minionId);
-                return false;
-            },
-            machineId -> {
-                boolean isReactivation = false;
-                ActivationKey ak = null;
-                ValueMap grains = new ValueMap(SALT_SERVICE.getGrains(minionId).orElseGet(HashMap::new));
-                Optional<String> managmentKeyLabel = getManagementKeyLabelFromGrains(grains);
-                if (managmentKeyLabel.isPresent()) {
-                    ak = ActivationKeyFactory.lookupByKey(managmentKeyLabel.get());
-                    if (ak != null && ak.getKickstartSession() != null) {
-                        ak.getKickstartSession().markComplete("Installation completed.");
-                    }
-                    if (ak == null) {
-                        LOG.info("Outdated Management Key defined for " + minionId + ": " + managmentKeyLabel.get());
-                    }
-                    else if (ak.getServer() == null) {
-                        LOG.error("Management Key is not a reactivation key: " + managmentKeyLabel.get());
-                    }
-                    else {
-                        isReactivation = true;
-                    }
+    private void registerMinion(String minionId, boolean isSaltSSH, Optional<Long> saltSSHProxyId,
+                                Optional<String> actKeyOverride, String machineId) {
+        if (!duplicateMinionNamePresent(minionId)) {
+            boolean isReactivation = false;
+            ActivationKey ak = null;
+            ValueMap grains = new ValueMap(SALT_SERVICE.getGrains(minionId).orElseGet(HashMap::new));
```

Comment:
Now I'm wondering what is the purpose of having startup grains. Because up until this point machine id is not used anywhere. Couldn't we just refactored registerMinion routing to check for machingID only after this grain call?

Or should we instead put all the necessary grains needed for registration to the startup grains and after the registration schedule hardware refresh to get the rest of the grains?

Or am I missing something?

Reply:
I have moved machin_id [check](https://github.com/uyuni-project/uyuni/pull/1833/files#diff-a112668bac573a2c5fb5981f604ffb99R119) outside of this function. Control will only move to this function if we will have a valid machine_id.